### ISSUE TITLE
Gray scale output instead of RGB

### ISSUE THREAD

**Issue by asmagen** [biomedia-mira/drop2#6]
Hello @bglocker and authors

I used `dropreg --mode2d -s CD3_test.png -t CD20_test.png -o CD20_to_CD3_test.png -l --llevels 4 4 1 2 2 1 --lsim 1 --lsampling 0.2 -n --nsim 1 --nffd 1000 --nlevels 4 4 1 2 2 1 2 2 1 --nlambda 0.5 --ocompose` to register two histopathology png RGB images and got a gray scale output. Why is that and what do I do to get the transformed image in the same RGB format as the original?
<img width="1668" alt="Screen Shot 2020-03-27 at 11 02 05 AM" src="https://user-images.githubusercontent.com/11299568/77769591-6ee23300-701a-11ea-8d42-d1d498c5cfce.png">

Also, when I do -s as image X and -t as image Y, would the program fit X to Y or vice versa? What do I get in the -o output?

**Comment by bglocker** [comment]
See #4 
drop2 does not support color images internally. It can load them, but they get converted to single-channel grayscale, and so is the output.

**Comment by asmagen** [comment]
Thanks, but can you clarify what steps do I need to get the deformation field and apply it on the original RGB image? I see something to do with landmarks which I don't have and not sure how to use.

Thanks

**Comment by bglocker** [comment]
The easiest way to do this is to use a tool (or Matlab or Python script) to separate your RGB images into three single-channel images, apply the saved transformation using drop2 on each individually, and put the results back together into an RGB image.

Note, you can use drop2 to apply a saved transformation by passing the transformation files as command line arguments using `--transform`, `--fx`, `--fy`, ...

**Comment by bglocker** [comment]
Just to add, if you use `--ocompose` the entire transformation (including the linear part) is stored in the displacement fields. So you only need to use `--fx` and `--fy` and pass on the x- and y-displacements to warp an image.

**Comment by asmagen** [comment]
Hello @bglocker 

I think it's done now, by applying the transformation to each channel and merging:
<img width="1139" alt="Screen Shot 2020-03-28 at 10 04 58 AM" src="https://user-images.githubusercontent.com/11299568/77824959-a0bbce00-70dc-11ea-9d8a-5d8f52814e71.png">

I used the following small piece of code to do that, but it's taking a lot of time and way too much memory such that it doesn't scale up to the big images I'm trying to analyze (15k x 15k pixels). So how does drop2 apply the transformation so quickly on the big size image in gray scale? And do you have any advice on improving/optimizing the code below or how to utilize the strategy you use on gray scale? You advised using drop2 --transform but it's not clear to me how exactly I can save the separate channel as a file to use an input and what inputs are requires (-s? -t?)

```
import os

import numpy as np
import nibabel
import matplotlib.pylab as plt
import skimage.color as sk_color
import SimpleITK as sitk
from scipy.interpolate import griddata
from skimage.transform import resize
import cv2 as cv

PATH_REGIST = '/sc/arion/projects/scRNATools/imaging/liver'
PATH_IMG_SRC = os.path.join(PATH_REGIST, 'Li63TCD8_cropped_mask.png')
PATH_IMG_REF = os.path.join(PATH_REGIST, 'Li63TCD3_cropped_mask.png')

PATH_DEF_X = os.path.join(PATH_REGIST, 'Li63T_CD20_transformed_field_x.nii.gz')
PATH_DEF_Y = os.path.join(PATH_REGIST, 'Li63T_CD20_transformed_field_y.nii.gz')

REF = plt.imread(PATH_IMG_REF)
SRC = plt.imread(PATH_IMG_SRC)

img_dx = sitk.ReadImage(PATH_DEF_X)
img_dy = sitk.ReadImage(PATH_DEF_Y)
spacing = img_dx.GetSpacing()
deform_x = sitk.GetArrayFromImage(img_dx)[0] / spacing[0]
deform_y = sitk.GetArrayFromImage(img_dy)[0] / spacing[1]
grid_y, grid_x = np.mgrid[0:deform_x.shape[0], 0:deform_x.shape[1]]

output = REF
output.shape
for level in (0,1,2):
    img_src = SRC[:,:,level]
    img_ref = REF[:,:,level]
    img_src_resized = resize(img_src, img_ref.shape)
    img_warp2 = griddata((grid_y.ravel(), grid_x.ravel()), img_src_resized.ravel(),(grid_y + deform_y, grid_x + deform_x), method='linear')
    output[:,:,level] = img_warp2

plt.imsave(os.path.join(PATH_REGIST, 'output_CD20.png'),output)
```

Thanks

**Comment by asmagen** [comment]
As a followup, although the transformation worked without any warnings/errors for the example tiles I used above, it did result in some kind of an error while running on a full image of ~11k x 14k pixels:
```
Traceback (most recent call last):
  File "<stdin>", line 5, in <module>
  File "/hpc/packages/minerva-centos7/py_packages/3.7/lib/python3.7/site-packages/scipy/interpolate/ndgriddata.py", line 222, in griddata
    rescale=rescale)
  File "interpnd.pyx", line 248, in scipy.interpolate.interpnd.LinearNDInterpolator.__init__
  File "qhull.pyx", line 1836, in scipy.spatial.qhull.Delaunay.__init__
  File "qhull.pyx", line 353, in scipy.spatial.qhull._Qhull.__init__
scipy.spatial.qhull.QhullError: QH6080 qhull error (qh_memalloc): insufficient memory to allocate short memory buffer (65536 bytes)

While executing:  | qhull d Qz Qt Qbb Q12 Qc
Options selected for Qhull 2015.2.r 2016/01/18:
  run-id 2113252558  delaunay  Qz-infinity-point  Qtriangulate  Qbbound-last
  Q12-no-wide-dup  Qcoplanar-keep  _pre-merge  _zero-centrum  Qinterior-keep
  Pgood  _max-width 1.7e+04  Error-roundoff 2.4e-11  _one-merge 1.7e-10
  Visible-distance 4.8e-11  U-coplanar-distance 4.8e-11  Width-outside 9.6e-11
  _wide-facet 2.9e-10
Last point added to hull was p27303443.  Last merge was #4738542.

At error exit:

Delaunay triangulation by the convex hull of 201763301 points in 3-d:

  Number of input sites and at-infinity: 13056075
  Total number of nearly incident points: 188707226
  Number of Delaunay regions: 0
  Number of non-simplicial Delaunay regions: 2489591

Statistics for:  | qhull d Qz Qt Qbb Q12 Qc

  Number of points processed: 13056075
  Number of hyperplanes created: 63907598
  Number of facets in hull: 22955992
  Number of distance tests for qhull: 50705539
  Number of distance tests for merging: 517448888
  Number of distance tests for checking: 0
  Number of merged facets: 4738542
  Maximum distance of merged point above facet: 6.1e-11 (0.3x)
  Maximum distance of merged vertex below facet: -5.6e-11 (0.3x)


precision problems (corrected unless 'Q0' or an error)
4738541 coplanar horizon facets for new vertices
```

Are you familiar with what could cause this?

Thank you

**Comment by asmagen** [comment]
You also advised using drop2 --transform but it's not clear to me how exactly I can save the separate channel as a file to use an input and what inputs are requires (-s? -t?)
